src/Hosting/Systemd/src/SystemdLifetime.cs

```diff
@@ -0,0 +1,90 @@
+    {
+        private readonly ManualResetEvent _shutdownBlock = new ManualResetEvent(false);
+        private CancellationTokenRegistration _applicationStartedRegistration;
+        private CancellationTokenRegistration _applicationStoppingRegistration;
+
+        public SystemdLifetime(IHostEnvironment environment, IHostApplicationLifetime applicationLifetime, ISystemdNotifier systemdNotifier, ILoggerFactory loggerFactory)
+        {
+            Environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            ApplicationLifetime = applicationLifetime ?? throw new ArgumentNullException(nameof(applicationLifetime));
+            SystemdNotifier = systemdNotifier ?? throw new ArgumentNullException(nameof(systemdNotifier));
+            Logger = loggerFactory.CreateLogger("Microsoft.Hosting.Lifetime");
+        }
+
+        private IHostApplicationLifetime ApplicationLifetime { get; }
+        private IHostEnvironment Environment { get; }
+        private ILogger Logger { get; }
+        private ISystemdNotifier SystemdNotifier { get; }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
+
+        public Task WaitForStartAsync(CancellationToken cancellationToken)
+        {
+            _applicationStartedRegistration = ApplicationLifetime.ApplicationStarted.Register(state =>
+            {
+                ((SystemdLifetime)state).OnApplicationStarted();
+            },
+            this);
+            _applicationStoppingRegistration = ApplicationLifetime.ApplicationStopping.Register(state =>
+            {
+                ((SystemdLifetime)state).OnApplicationStopping();
+            },
+            this);
+
+            // systemd sends SIGTERM to stop the service.
+            AppDomain.CurrentDomain.ProcessExit += OnProcessExit;
+
+            return Task.CompletedTask;
+        }
+
+        private void OnApplicationStarted()
+        {
+            Logger.LogInformation("Application started. Hosting environment: {envName}; Content root path: {contentRoot}",
+                Environment.EnvironmentName, Environment.ContentRootPath);
+
+            SystemdNotifier.Notify(ServiceState.Ready);
+        }
+
+        private void OnApplicationStopping()
+        {
+            Logger.LogInformation("Application is shutting down...");
+
+            SystemdNotifier.Notify(ServiceState.Stopping);
+        }
+
+        private void OnProcessExit(object sender, EventArgs e)
+        {
+            ApplicationLifetime.StopApplication();
+
+            _shutdownBlock.WaitOne();
+
+            // On Linux if the shutdown is triggered by SIGTERM then that's signaled with the 143 exit code.
+            // Suppress that since we shut down gracefully. https://github.com/aspnet/AspNetCore/issues/6526
+            System.Environment.ExitCode = 0;
+        }
+
+        public void Dispose()
+        {
+            _shutdownBlock.Set();
```

Comment:
What happens if SystemdLifetime isn't disposed? Wouldn't OnProcessExit hang?

Reply:
OnProcessExit calls ApplicationLifetime.StopApplication which should dispose the DI container and therefore this. That said, it might not be a bad idea to timeout _shutdownBlock.WaitOne() and throw if it does timeout which a message explaining a SystemdLifetime didn't get disposed for some reason.

I'm guessing, thought I'm not sure, that the OnProcessExit call itself will eventually timeout (even if it's just because the process terminates) so shutdown doesn't actually hang forever. If that's true, we'd want to make sure we throw prior to that timeout. And we'd also want to make sure that exceptions thrown from OnProcessExit are easily observed.

In practice, I doubt this will come up much though, so I'm not too worried about it.

Reply:
Yes it would hang. We have another thread on this: https://github.com/aspnet/Extensions/issues/1363